timelock-impl/src/main/java/com/palantir/atlasdb/timelock/adjudicate/FeedbackProvider.java

```diff
@@ -0,0 +1,139 @@
+/*
+ * (c) Copyright 2020 Palantir Technologies Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.atlasdb.timelock.adjudicate;
+
+import java.util.List;
+import java.util.Map;
+import java.util.UUID;
+import java.util.function.Function;
+import java.util.stream.Stream;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.collect.Lists;
+import com.google.common.collect.Maps;
+import com.palantir.common.streams.KeyedStream;
+import com.palantir.timelock.feedback.ConjureTimeLockClientFeedback;
+import com.palantir.timelock.feedback.EndpointStatistics;
+
+public final class FeedbackProvider {
+
+    private FeedbackProvider() {
+        // no op
+    }
+
+    public static HealthStatus getTimeLockHealthStatus(TimeLockClientFeedbackSink timeLockClientFeedbackSink) {
+        List<ConjureTimeLockClientFeedback> trackedFeedbackReports =
+                timeLockClientFeedbackSink.getTrackedFeedbackReports();
+
+        Map<String, Map<UUID, List<ConjureTimeLockClientFeedback>>> organizedFeedback =
+                organizeFeedbackReports(trackedFeedbackReports);
+
+        return healthStateOfTimeLock(organizedFeedback);
+    }
+
+    private static Map<String, Map<UUID, List<ConjureTimeLockClientFeedback>>> organizeFeedbackReports(
+            List<ConjureTimeLockClientFeedback> trackedFeedbackReports) {
+        Map<String, Map<UUID, List<ConjureTimeLockClientFeedback>>> organizedFeedback = Maps.newHashMap();
+
+        for (ConjureTimeLockClientFeedback feedback : trackedFeedbackReports) {
+            Map<UUID, List<ConjureTimeLockClientFeedback>> nodesForService = organizedFeedback.computeIfAbsent(
+                    feedback.getServiceName(), service -> Maps.newHashMap());
+
+            List<ConjureTimeLockClientFeedback> feedbackForNode = nodesForService
+                    .computeIfAbsent(feedback.getNodeId(), nodeId -> Lists.newLinkedList());
```

Comment:
`ArrayList` is normally preferred here, because of locality (i.e. the entries are stored close together in memory, which is useful because  we will end up accessing them sequentially later on).